### ISSUE TITLE
how to generate balanced batches for multilingual translation? 

### ISSUE THREAD

**Issue by CharizardAcademy** [pytorch/fairseq#890]
I want to use the basic transformer architecture to do multilingual translation, that is, instead of having separate encoders and decoders for each language pair, I want one encoder and decoder for all language pairs, thus the batches should be balanced for all language pairs according to the corpus size, in order to prevent overfitting one language pair. But I find it hard to implement, since the base transformer only reads in one dataset and generates batches according to sequence length. If I have a dataset that contains all languages pairs, the batch generator still cannot guarantee to generate balanced batches. Is there a way to do this? Thanks

**Comment by pipibjc** [comment]
One easy way is to keep current way to load the dataset, and we re-weigh the loss of batches from each language pair as the ratio of original data. At `train_step` of `MultilingualTranslationTask`, we [multiply the loss ](https://github.com/pytorch/fairseq/blob/e4335001d829ae806167af09aeac2b42c95ee764/fairseq/tasks/multilingual_translation.py#L280)by the weight of the language pair.
The weight can be passed in some parameter, or we can compute it on the fly in `load_dataset`.

**Comment by lematt1991** [comment]
Closing due to inactivity.
